kong/router/atc_compat.lua

```diff
@@ -0,0 +1,586 @@
+
+  local priority =  bor(match_weight,
+                        plain_host_only and lshift(0x01ULL, 60) or 0,
+                        regex_url and lshift(0x01ULL, 51) or 0,
+                        headers_count,
+                        regex_priority,
+                        max_length)
+
+  return priority
+end
+
+
+function _M.new(routes)
+  if type(routes) ~= "table" then
+    return error("expected arg #1 routes to be a table")
+  end
+
+  local s = get_schema()
+  local inst = router.new(s)
+
+  local router = setmetatable({
+    schema = s,
+    router = inst,
+    routes = {},
+    services = {},
+    fields = {},
+  }, _MT)
+
+  for _, r in ipairs(routes) do
+    local route = r.route
+    local route_id = route.id
+    router.routes[route_id] = route
+    router.services[route_id] = r.service
+
+    if kong.configuration.router_flavor == "traditional_compatible" then
+      assert(inst:add_matcher(route_priority(route), route_id, get_atc(route)))
+
+    else
+      local atc = route.atc
+
+      local gen = gen_for_field("net.protocol", OP_EQUAL, route.protocols)
+      if gen then
+        atc = atc .. " && " .. gen
+      end
+
+      assert(inst:add_matcher(route.priority, route_id, atc))
+    end
+
+    router.fields = inst:get_fields()
+  end
+
+  return router
+end
+
+
+local function sanitize_uri_postfix(uri_postfix)
+  if not uri_postfix or uri_postfix == "" then
+    return uri_postfix
+  end
+
+  if uri_postfix == "." or uri_postfix == ".." then
+    return ""
+  end
+
+  if sub(uri_postfix, 1, 2) == "./" then
+    return sub(uri_postfix, 3)
+  end
+
+  if sub(uri_postfix, 1, 3) == "../" then
+    return sub(uri_postfix, 4)
+  end
+
+  return uri_postfix
+end
+
+
+function _M:select(req_method, req_uri, req_host, req_scheme,
+                   src_ip, src_port,
+                   dst_ip, dst_port,
+                   sni, req_headers)
+  if req_method and type(req_method) ~= "string" then
+    error("method must be a string", 2)
+  end
+  if req_uri and type(req_uri) ~= "string" then
+    error("uri must be a string", 2)
+  end
+  if req_host and type(req_host) ~= "string" then
+    error("host must be a string", 2)
+  end
+  if req_scheme and type(req_scheme) ~= "string" then
+    error("scheme must be a string", 2)
+  end
+  if src_ip and type(src_ip) ~= "string" then
+    error("src_ip must be a string", 2)
+  end
+  if src_port and type(src_port) ~= "number" then
+    error("src_port must be a number", 2)
+  end
+  if dst_ip and type(dst_ip) ~= "string" then
+    error("dst_ip must be a string", 2)
+  end
+  if dst_port and type(dst_port) ~= "number" then
+    error("dst_port must be a number", 2)
+  end
+  if sni and type(sni) ~= "string" then
+    error("sni must be a string", 2)
+  end
+  if req_headers and type(req_headers) ~= "table" then
+    error("headers must be a table", 2)
+  end
+
+  local c = context.new(self.schema)
+
+  for _, field in ipairs(self.fields) do
```

Comment:
it seems headers are the only variables that we fetch lazily. Could we just unconditionally add all other fields here?